File: frontend/src/apis/enrollment.ts
import client, { handleEnvelope, handlePaginatedEnvelope, ApiError } from '@/apis/client'

export type Semester = 'fall' | 'spring' | 'summer' | 'winter' | 'interim' | 'first' | 'second' | 'third' | 'fourth' | 'fifth' | 'sixth' | 'seventh' | 'eighth'

export interface Enrollment {
  id: string
  student_id: string
  student_name?: string
  student_email?: string
  institution_id?: string
  course_id: string
  course_name?: string
  lecturer_id?: string
  semester: Semester
  year?: number
  status: 'active' | 'completed' | 'dropped' | 'pending'
  created_at: string
  updated_at: string
}

export interface EnrollmentCreate {
  student_id: string
  course_id: string
  semester: Semester
  year?: number
}

export interface EnrollmentUpdate {
  status?: Enrollment['status']
}

export interface EnrollmentCheckResponse {
  enrolled: boolean
  enrollment?: Enrollment
}

export interface EnrollmentListParams {
  page?: number
  per_page?: number
  search?: string
  student_id?: string
  course_id?: string
  lecturer_id?: string
  status?: Enrollment['status']
  semester?: Semester
  year?: number
}

export interface EnrollmentListResponse {
  items: Enrollment[]
  pagination: {
    page: number
    per_page: number
    total: number
    pages: number
    has_next: boolean
    has_prev: boolean
  }
}

/** Extract the most useful message from an axios/API error */
function extractMessage(error: unknown, fallback: string): string {
  if (error instanceof ApiError) return error.message
  const e = error as any
  // FastAPI HTTPException detail
  const detail = e?.response?.data?.detail
  if (typeof detail === 'string') return detail
  // Our custom Response error field
  const apiError = e?.response?.data?.error
  if (typeof apiError === 'string') return apiError
  if (e?.message) return e.message
  return fallback
}

export async function listEnrollment(params: EnrollmentListParams = {}): Promise<EnrollmentListResponse> {
  try {
    const response = await client.get(`/academic/enrollment/`, { params })
    return handlePaginatedEnvelope<Enrollment>(response)
  } catch (error) {
    throw new ApiError(extractMessage(error, 'Failed to fetch enrollments'))
  }
}

export async function getEnrollment(id: string): Promise<Enrollment> {
  try {
    const response = await client.get(`/academic/enrollment/${id}`)
    return handleEnvelope<Enrollment>(response)
  } catch (error) {
    throw new ApiError(extractMessage(error, 'Failed to fetch enrollment'))
  }
}

export async function enrollStudent(data: EnrollmentCreate): Promise<Enrollment> {
  try {
    const response = await client.post(`/academic/enrollment/`, data)
    return handleEnvelope<Enrollment>(response)
  } catch (error) {
    throw new ApiError(extractMessage(error, 'Failed to enroll student'))
  }
}

export async function updateEnrollment(id: string, data: EnrollmentUpdate): Promise<Enrollment> {
  try {
    const response = await client.put(`/academic/enrollment/${id}`, data)
    return handleEnvelope<Enrollment>(response)
  } catch (error) {
    throw new ApiError(extractMessage(error, 'Failed to update enrollment'))
  }
}

export async function removeEnrollment(id: string): Promise<void> {
  try {
    await client.delete(`/academic/enrollment/${id}`)
  } catch (error) {
    throw new ApiError(extractMessage(error, 'Failed to remove enrollment'))
  }
}

export async function getCourseEnrollment(courseId: string, params: Omit<EnrollmentListParams, 'course_id'> = {}): Promise<EnrollmentListResponse> {
  return listEnrollment({ ...params, course_id: courseId })
}

export async function getStudentEnrollment(studentId: string, params: Omit<EnrollmentListParams, 'student_id'> = {}): Promise<EnrollmentListResponse> {
  return listEnrollment({ ...params, student_id: studentId })
}

export async function getLecturerEnrollment(lecturerId: string, params: Omit<EnrollmentListParams, 'lecturer_id'> = {}): Promise<EnrollmentListResponse> {
  return listEnrollment({ ...params, lecturer_id: lecturerId })
}

export async function bulkEnroll(enrollments: EnrollmentCreate[]): Promise<Enrollment[]> {
  try {
    const response = await client.post(`/academic/enrollment/bulk`, { enrollments })
    return handleEnvelope<Enrollment[]>(response)
  } catch (error) {
    throw new ApiError(extractMessage(error, 'Failed to bulk enroll students'))
  }
}

export async function checkEnrollment(studentId: string, courseId: string): Promise<EnrollmentCheckResponse> {
  try {
    const response = await client.get(`/academic/enrollment/check`, {
      params: { student_id: studentId, course_id: courseId }
    })
    return handleEnvelope<EnrollmentCheckResponse>(response)
  } catch (error) {
    throw new ApiError(extractMessage(error, 'Failed to check enrollment'))
  }
}

File: backend/schemas/academic/enrollment.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class EnrollmentStatus(str, Enum):
    ACTIVE = "active"
    COMPLETED = "completed"
    DROPPED = "dropped"
    PENDING = "pending"


class Semester(str, Enum):
    FIRST = "first"
    SECOND = "second"
    THIRD = "third"
    FOURTH = "fourth"
    FIFTH = "fifth"
    SIXTH = "sixth"
    SEVENTH = "seventh"
    EIGHTH = "eighth"
    FALL = "fall"
    SPRING = "spring"
    SUMMER = "summer"
    WINTER = "winter"
    INTERIM = "interim"


class EnrollmentBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    student_id: UUID
    course_id: UUID
    semester: Semester
    year: Optional[int] = Field(None, ge=2020, le=2100)


class EnrollmentCreate(EnrollmentBase):
    pass


class EnrollmentUpdate(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    status: Optional[EnrollmentStatus] = None


class EnrollmentRead(EnrollmentBase):
    id: UUID
    student_name: Optional[str] = None
    student_email: Optional[str] = None
    institution_id: Optional[str] = None
    course_name: Optional[str] = None
    lecturer_id: Optional[UUID] = None
    status: EnrollmentStatus
    created_at: datetime
    updated_at: datetime


# Alias for API responses
EnrollmentResponse = EnrollmentRead


class EnrollmentListParams(BaseModel):
    """Query parameters for listing enrollments."""
    course_id: Optional[UUID] = None
    student_id: Optional[UUID] = None
    status: Optional[EnrollmentStatus] = None
    semester: Optional[Semester] = None
    year: Optional[int] = None
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)


class BulkEnrollmentRequest(BaseModel):
    """Request body for bulk enrollment."""
    course_id: UUID
    student_ids: List[UUID]
    semester: Semester
    year: Optional[int] = Field(None, ge=2020, le=2100)


class EnrollmentCheckRequest(BaseModel):
    """Request to check if a student is enrolled in a course."""
    student_id: UUID
    course_id: UUID


class EnrollmentCheckResponse(BaseModel):
    """Response for enrollment check."""
    is_enrolled: bool
    enrollment: Optional[EnrollmentRead] = None


class EnrollmentListResponse(BaseModel):
    """Paginated list of enrollments."""
    data: List[EnrollmentRead]
    total: int
    page: int
    page_size: int


class EnrollmentDelete(BaseModel):
    id: UUID
